### ISSUE TITLE
Why is NlSocket::send/recv taking `&mut self`?

### ISSUE THREAD

**Issue by vorner** [jbaublitz/neli#41]
Hello

I'm wondering about why does the send/recv method (and send_nl/recv_nl) taking *mutable* self? Looking at the code, there doesn't seem to be a reason to do so.

My motivation for asking is this. There's the NFQUEUE protocol, which allows implementing part of firewall in userspace. The kernel sends packets to userspace over netlink socket and the program sends verdicts back. It is quite common to receive the packets in one thread and send them back in another. This works without any locking in the userspace part ‒ I suppose recv/send lock as necessary in kernel. So it would be better to do so with just `&self`.

**Comment by jbaublitz** [comment]
Hi @vorner, the reason for this is that there should never be two threads reading from a socket at the same time just as there should never be two threads writing to the socket at the same time. This would cause corruption. Mutable references protect against this issue. It may make sense to implement a method that splits a socket object into one read-only and one write-only handle as I've seen some other libraries do for situations like this. Would that be helpful?

**Comment by vorner** [comment]
Does it produce corruption? I thought netlink was message-based, therefore send/recv over it was atomic and could be done from multiple threads even the way you describe.

If that's so, it would still be great to have the read/write half possibility.

**Comment by jbaublitz** [comment]
I've seen no evidence that sugggests that netlink is atomic, and I've looked again because of this issue. However, that does not mean that it is not atomic, simply that I was unable to find any data specifically on netlink. I will say that I'm wary about relaxing this protection without direct evidence indicating that this is safe. If you can provide me with documentation or netlink kernel code indicating that this is indeed safe, I would be happy to consider it for the v0.5.0 release, but as it stands I haven't seen anything in my research that supports this is 100% safe.

**Comment by jbaublitz** [comment]
> A single read is atomic, but that might be a bit misleading. Messages will not always be received in their entirety in one read. One simple example is when the message is large, and the receiving buffer is small.

Any data that's available in a given message that is not read will be truncated. I view this as orthogonal to the discussion of whether this is atomic or not. I can see that this may be a side effect of atomic reads but I am not convinced that this is an indicator without more concrete evidence. Again, if that's provided, I'm happy to reconsider and change that for v0.5.0. I will also make a note of looking into this myself to see if I can find some kind of locking in the kernel that would provide atomic reads and writes. 

The only example above that makes me feel a bit better about the idea of dropping `mut` is zmq. That is indeed an example of a socket with atomic operations. In the case of nix, it is merely a wrapper for the syscalls which could still cause corruption on certain socket types if a file descriptor of a TCP socket is provided for example. My goal is to avoid corruption while I think nix is more aiming to provide a low level but more idiomatic interface without unsafe keywords to linux syscalls. As for the standard library example, this is again a slightly different case where the operation is guaranteed to be atomic at the hardware level. Socket operations can't make that guarantee without kernel support for locking or some other method of concurrent protection.

I will look and see how ZMQ sockets are created as it very well may be that I'm wrong here and raw sockets as are used with netlink can be considered atomic but I need to do more research or see more convincing data on this before I'm willing to incorporate that assumption into neli.

**Comment by vorner** [comment]
I believe that the atomicity of recv/send itself and the fact that no message can span more than one read is an indication that this is safe from the kernel side. There's the question about what to do with sequence numbers and such, though, which is unclear.

Anyway, for my use cases I believe being able to do a split into read-half and write-half does suffice.

**Comment by jbaublitz** [comment]
I've poked around a bit and I've found documentation that message based reads and writes are indeed atomic in the Linux kernel. This is new information for me, and I appreciate you both pointing me in the right direction. I'll target immutable receivers in `send()` and `recv()` for v0.5.0. I'll probably still do some testing to verify, but this appears to be so well trusted for message-based sockets that I found mentions of UDP sockets sometimes having a pool of readers all reading from the same socket, each handling a single complete message without synchronization. Thanks again!
